File: scripts/export-pyrit-digest.py
#!/usr/bin/env python3
"""Emit a precompiled, Python-`re`-verified rule digest for downstream scorers.

Why this exists: microsoft/PyRIT#1893 added an ATR scorer that imported `pyatr`,
and the maintainer reverted it a day later (#2410) because it made PyRIT
responsible for a third-party package's release cadence. He was right, and the
published wheel proves it — `pyatr` 0.2.7 (2026-07-10) bakes in 713 rules while
this repo is well past that. A consumer that vendors an engine inherits our
release schedule; a consumer that fetches a pinned data artifact does not.

So this emits data, not code: `{rule_id: {patterns, category, severity, ...}}`,
consumed by a `RegexScorer` subclass on the PyRIT side with stdlib `re` and no
new dependency.

Two honesty properties matter more than the file itself:

1. Every pattern is `re.compile()`d here, at build time. A pattern that does not
   compile never reaches the artifact, and the reason is recorded by rule id.
   The failure mode this avoids is Microsoft's own weekly ATR sync in
   agent-governance-toolkit, which runs without `--strict-regex` and skips
   invalid patterns with a warning — so an unknown number of rules are dropped
   from a shipped product and nobody downstream can tell which.

2. Exclusions are diffed against a checked-in baseline and the build FAILS when
   the set changes. Known breakage stays visible; new breakage stops the build.
   Silence is the thing being engineered out.

Note the engine split, because quoting one number for both is wrong: Python `re`
is a backtracking engine and accepts lookahead, backreferences and fixed-width
lookbehind. RE2 (Go, Rust) accepts none of them. A pattern excluded here is a
strictly worse case than a pattern excluded from an RE2 export.

  python3 scripts/export-pyrit-digest.py
  python3 scripts/export-pyrit-digest.py --strict          # any exclusion fails
  python3 scripts/export-pyrit-digest.py --update-baseline
"""
import argparse
import glob
import json
import os
import re
import subprocess
import sys

import yaml


def _load_sigma_helpers():
    """Reuse the regex helpers from generate-sigma.py.

    Loaded by path because the filename is hyphenated and therefore not a legal
    module name. Reused rather than reimplemented on purpose: that file already
    owns the escape/character-class-aware regex walk and the inline-flag
    handling, and a second copy of that logic would drift from it silently.
    """
    import importlib.util

    path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "generate-sigma.py")
    spec = importlib.util.spec_from_file_location("atr_generate_sigma", path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"cannot load {path}")
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


_sigma = _load_sigma_helpers()
RULES_GLOB = _sigma.RULES_GLOB
strip_leading_i_flag = _sigma.strip_leading_i_flag

DEFAULT_OUT = "data/pyrit/atr-rule-digest.json"
DEFAULT_BASELINE = "data/pyrit/atr-rule-digest-exclusions.json"
DIGEST_SCHEMA_VERSION = "1"

# Fields a PyRIT scorer can actually evaluate. It is handed one MessagePiece —
# a single string — so a condition bound to anything else is not "supported but
# untested", it is unreachable. Recording that per rule rather than dropping it
# silently is the same discipline as reporting unevaluated rules instead of
# counting them clean.
TEXT_FIELDS = frozenset(
    {
        "content",
        "user_input",
        "tool_response",
        "agent_output",
        "tool_args",
        "tool_description",
        "tool_input",
    }
)


def repo_root() -> str:
    return os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


def head_commit() -> str:
    try:
        out = subprocess.run(
            ["git", "rev-parse", "HEAD"],
            cwd=repo_root(),
            capture_output=True,
            text=True,
            check=True,
        )
        return out.stdout.strip()
    except (subprocess.CalledProcessError, FileNotFoundError):
        return "unknown"


def compile_failure(pattern: str) -> str | None:
    """Return the reason `pattern` will not compile under Python `re`, or None."""
    try:
        re.compile(pattern)
    except re.error as exc:
        return str(exc).split(" at position")[0]
    return None


def build(rules_glob: str) -> tuple[dict, list[dict], int]:
    rules: dict[str, dict] = {}
    exclusions: list[dict] = []
    seen = 0

    for path in sorted(glob.glob(rules_glob, recursive=True)):
        try:
            with open(path, encoding="utf-8") as fh:
                rule = yaml.safe_load(fh)
        except (OSError, yaml.YAMLError) as exc:
            exclusions.append({"rule": os.path.basename(path), "reason": "unreadable", "detail": str(exc)[:160]})
            continue
        if not isinstance(rule, dict):
            continue

        rid = rule.get("id")
        if not rid:
            continue
        seen += 1

        detection = rule.get("detection") or {}
        combinator = str(detection.get("condition") or "").strip().lower()
        if combinator != "any":
            # RegexScorer reports a hit when ANY named pattern matches. A rule
            # that requires all of its conditions cannot be represented without
            # inventing matches it never claimed, so it is left out rather than
            # approximated.
            exclusions.append({"rule": rid, "reason": "condition-not-any", "detail": combinator or "(absent)"})
            continue

        patterns: list[str] = []
        skipped_fields: set[str] = set()
        failed: list[dict] = []

        for cond in detection.get("conditions") or []:
            if not isinstance(cond, dict) or cond.get("operator") != "regex":
                continue
            field = str(cond.get("field"))
            value = cond.get("value")
            if not isinstance(value, str) or not value:
                continue
            if field not in TEXT_FIELDS:
                skipped_fields.add(field)
                continue
            reason = compile_failure(value)
            if reason:
                failed.append({"reason": reason, "pattern": value[:120]})
                continue
            patterns.append(value)

        for f in failed:
            exclusions.append({"rule": rid, "reason": "pattern-uncompilable", "detail": f["reason"]})

        if not patterns:
            exclusions.append(
                {
                    "rule": rid,
                    "reason": "no-usable-pattern",
                    "detail": (
                        f"{len(failed)} uncompilable, "
                        f"{len(skipped_fields)} non-text field(s): {','.join(sorted(skipped_fields)) or 'none'}"
                    ),
                }
            )
            continue

        # `(?i)` is only legal at the start of a Python pattern, so a consumer
        # that concatenates patterns would break. The flag is lifted here and
        # declared once, which keeps every pattern independently compilable.
        lifted, any_ci = [], False
        for p in patterns:
            body, had = strip_leading_i_flag(p)
            any_ci = any_ci or had
            lifted.append(body if had else p)

        rules[rid] = {
            "title": rule.get("title") or "",
            "category": os.path.basename(os.path.dirname(path)),
            "severity": rule.get("severity") or "unknown",
            "maturity": rule.get("maturity") or "unknown",
            "status": rule.get("status") or "unknown",
            "ignore_case": any_ci,
            "patterns": lifted if any_ci else patterns,
            "unreachable_fields": sorted(skipped_fields),
        }

    return rules, exclusions, seen


def exclusion_key(entries: list[dict]) -> list[list[str]]:
    """Stable, order-independent shape for baseline comparison."""
    return sorted([e["rule"], e["reason"]] for e in entries)


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--out", default=DEFAULT_OUT)
    ap.add_argument("--baseline", default=DEFAULT_BASELINE)
    ap.add_argument(
        "--root",
        default=None,
        help="repository root to read rules from; defaults to this script's repo. Exists so the tests can run the real script against a fixture tree instead of the live corpus.",
    )
    ap.add_argument("--strict", action="store_true", help="fail if ANY rule is excluded")
    ap.add_argument("--update-baseline", action="store_true")
    args = ap.parse_args()

    os.chdir(args.root or repo_root())
    rules, exclusions, total_seen = build(RULES_GLOB)
    # A rule can be both included and carry an exclusion (one bad pattern among
    # several good ones), so these two counts deliberately overlap. Report the
    # rules that produced NOTHING separately from the total exclusion entries.
    dropped = len({e["rule"] for e in exclusions if e["reason"] in ("no-usable-pattern", "condition-not-any", "unreadable")})

    digest = {
        "$comment": "AUTO by scripts/export-pyrit-digest.py — do not edit by hand.",
        "schema_version": DIGEST_SCHEMA_VERSION,
        "engine": "python-re",
        "source_commit": head_commit(),
        "rules_total": total_seen,
        "rules_included": len(rules),
        "rules_dropped": dropped,
        "exclusion_entries": len(exclusions),
        "exclusions": sorted(exclusions, key=lambda e: (e["rule"], e["reason"])),
        "rules": dict(sorted(rules.items())),
    }

    os.makedirs(os.path.dirname(args.out), exist_ok=True)
    with open(args.out, "w", encoding="utf-8") as fh:
        json.dump(digest, fh, ensure_ascii=False, indent=2)
        fh.write("\n")

    current = exclusion_key(exclusions)
    print(
        f"wrote {args.out} — {len(rules)}/{total_seen} rules included, "
        f"{dropped} dropped entirely, {len(exclusions)} exclusion entries"
    )

    if args.update_baseline:
        os.makedirs(os.path.dirname(args.baseline), exist_ok=True)
        with open(args.baseline, "w", encoding="utf-8") as fh:
            json.dump(
                {
                    "$comment": "Known exclusions from the PyRIT digest. The build fails when this set changes; update deliberately.",
                    "exclusions": current,
                },
                fh,
                ensure_ascii=False,
                indent=2,
            )
            fh.write("\n")
        print(f"baseline updated — {len(current)} known exclusions")
        return 0

    if args.strict and exclusions:
        print(f"STRICT: {len(exclusions)} exclusion(s); every rule must be representable", file=sys.stderr)
        return 1

    if not os.path.exists(args.baseline):
        print(f"no baseline at {args.baseline} — run with --update-baseline once to record the current set", file=sys.stderr)
        return 1

    with open(args.baseline, encoding="utf-8") as fh:
        known = [list(x) for x in json.load(fh).get("exclusions", [])]

    added = [x for x in current if x not in known]
    removed = [x for x in known if x not in current]
    if added or removed:
        print("\nEXCLUSION SET CHANGED — failing rather than skipping quietly.", file=sys.stderr)
        for rule, reason in added:
            print(f"  + {rule}  {reason}", file=sys.stderr)
        for rule, reason in removed:
            print(f"  - {rule}  {reason}  (fixed — rerun with --update-baseline)", file=sys.stderr)
        return 1

    print(f"exclusions match baseline ({len(known)})")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/export-pyrit-digest.ts
#!/usr/bin/env node
/**
 * Emit a precompiled rule digest for PyRIT's AgentThreatRulesScorer.
 *
 * PyRIT reverted the first version of that scorer (microsoft/PyRIT#2410) because
 * it added `pyatr` as a dependency. The replacement subclasses PyRIT's existing
 * `RegexScorer`, which takes `dict[name, regex]` and compiles each with a plain
 * `re.compile(pattern)` -- no flags, no engine. This file produces that dict.
 *
 * TWO THINGS ARE BAKED IN HERE RATHER THAN LEFT TO THE CONSUMER
 *
 * Case sensitivity. The ATR engine compiles case-insensitively unless a
 * condition sets `case_sensitive`, but `RegexScorer` passes no flags. Measured
 * on the current rule set, 1,082 of 3,322 regex conditions (32.6%) carry no
 * `(?i)` prefix and would silently become case-sensitive on the other side. Each
 * emitted pattern therefore carries its own `(?i)` unless the condition opted
 * out, so the digest means the same thing wherever it is compiled.
 *
 * Astral escapes. `\u{...}` is JavaScript-only and does not compile under Python
 * `re`; the rules use literal characters for exactly this reason (see
 * needsUnicodeFlag in src/engine.ts). Nothing to translate, but patterns are
 * still test-compiled below and anything that fails is dropped with its reason
 * recorded rather than shipped broken.
 *
 * WHAT IS NOT REPRESENTABLE
 *
 * `RegexScorer` is a pure OR across every pattern, so a rule whose conditions
 * must ALL hold cannot be expressed. Four rules use `condition: all`; they are
 * excluded and listed in the digest's `excluded` block, because a rule silently
 * downgraded from AND to OR is a false-positive generator, not a rule.
 *
 * Field routing is likewise absent: the scorer sees one string, so a condition
 * written against `tool_args` is applied to whatever it is given. That is a
 * property of the consumer, and the digest records each pattern's field so the
 * caller can filter if it ever gains the context to.
 *
 * Usage:
 *   npx tsx scripts/export-pyrit-digest.ts --out data/pyrit-digest.json
 */
import { readFileSync, readdirSync, statSync, writeFileSync } from 'node:fs';
import { join, resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { execFileSync } from 'node:child_process';
import { load as parseYaml } from 'js-yaml';

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const RULES = join(ROOT, 'rules');

interface Emitted {
  readonly rule_id: string;
  readonly pattern: string;
  readonly field: string;
  readonly category: string;
  readonly severity: string;
}

function ruleFiles(dir: string): string[] {
  const out: string[] = [];
  for (const e of readdirSync(dir)) {
    const f = join(dir, e);
    if (statSync(f).isDirectory()) out.push(...ruleFiles(f));
    else if (e.endsWith('.yaml') || e.endsWith('.yml')) out.push(f);
  }
  return out;
}

function main(argv: readonly string[]): number {
  const outIdx = argv.indexOf('--out');
  const out = outIdx >= 0 ? argv[outIdx + 1] : undefined;
  const fieldIdx = argv.indexOf('--fields');
  // Which fields the consumer can actually present. A scorer that sees a model
  // response can offer agent_output and content and nothing else; emitting a
  // condition written against tool_args means applying it to prose it was never
  // meant for. Measured on 600 ordinary conversations: the unfiltered digest
  // flags 23.5% where the engine flags 0.8%, and 122 of the 141 extra come from
  // one rule whose conditions name tool_name and tool_args.
  const fields = new Set(
    (fieldIdx >= 0 ? argv[fieldIdx + 1] : 'agent_output,content').split(',').map((f) => f.trim()),
  );
  if (!out) {
    console.error('usage: export-pyrit-digest.ts --out <path> [--fields agent_output,content]');
    return 2;
  }

  const patterns: Record<string, string> = {};
  const meta: Record<string, Emitted> = {};
  const excluded: Record<string, string> = {};
  let rulesSeen = 0;
  let rulesEmitted = 0;
  let outOfScope = 0;

  for (const file of ruleFiles(RULES).sort()) {
    let doc: unknown;
    try {
      doc = parseYaml(readFileSync(file, 'utf-8'));
    } catch (e) {
      excluded[file] = `unparseable: ${String(e).slice(0, 60)}`;
      continue;
    }
    const r = doc as Record<string, any>;
    if (typeof r?.id !== 'string') continue;

    const status = String(r.status ?? '');
    const maturity = String(r.maturity ?? '');
    // Inert rules never fire in the engine either.
    if (status === 'draft' || status === 'deprecated' || maturity === 'deprecated') continue;
    rulesSeen += 1;

    const det = r.detection ?? {};
    if (String(det.condition ?? 'any') === 'all') {
      excluded[r.id] = 'condition: all -- RegexScorer is a pure OR and cannot express it';
      continue;
    }

    const category = String(r.tags?.category ?? '');
    const severity = String(r.severity ?? '').toLowerCase();
    let emittedForRule = 0;

    for (const [idx, cond] of (det.conditions ?? []).entries()) {
      if (String(cond?.operator ?? '') !== 'regex') continue;
      const raw = String(cond.value ?? '');
      if (!raw) continue;
      const field = String(cond.field ?? 'content');
      if (!fields.has(field)) {
        outOfScope += 1;
        continue;
      }

      // Bake in the flag the engine would have applied.
      const alreadyInline = /^\(\?[imsx]+\)/.test(raw);
      const pattern = cond.case_sensitive || alreadyInline ? raw : `(?i)${raw}`;

      // Ship nothing that does not compile where it will be used.
      try {
        execFileSync('python3', ['-c', 'import re,sys; re.compile(sys.argv[1])', pattern], {
          stdio: 'pipe',
        });
      } catch {
        excluded[`${r.id}#${idx}`] = 'does not compile under Python re';
        continue;
      }

      const name = `${r.id}#${idx}`;
      patterns[name] = pattern;
      meta[name] = {
        rule_id: r.id,
        pattern,
        field: String(cond.field ?? 'content'),
        category,
        severity,
      };
      emittedForRule += 1;
    }
    if (emittedForRule > 0) rulesEmitted += 1;
  }

  let commit = 'unknown';
  try {
    commit = execFileSync('git', ['rev-parse', 'HEAD'], { cwd: ROOT, encoding: 'utf-8' }).trim();
  } catch {
    /* not a checkout */
  }
  const version = JSON.parse(readFileSync(join(ROOT, 'package.json'), 'utf-8')).version;

  const digest = {
    _comment:
      'Precompiled ATR rule digest for consumers that cannot take a dependency on an ATR engine. ' +
      'Each pattern carries its own inline flags and compiles under Python re as-is. ' +
      'Generated by scripts/export-pyrit-digest.ts -- do not hand-edit.',
    atr_version: version,
    atr_commit: commit,
    fields_in_scope: [...fields].sort(),
    conditions_out_of_scope: outOfScope,
    rules_seen: rulesSeen,
    rules_emitted: rulesEmitted,
    pattern_count: Object.keys(patterns).length,
    patterns,
    meta,
    excluded,
  };
  writeFileSync(out, JSON.stringify(digest, null, 2) + '\n');
  console.log(
    `[pyrit-digest] ${rulesEmitted}/${rulesSeen} rules, ${Object.keys(patterns).length} patterns, ` +
      `${Object.keys(excluded).length} excluded, ${outOfScope} out of field scope -> ${out}`,
  );
  return 0;
}

const DIRECT =
  process.argv[1] !== undefined && resolve(process.argv[1]) === resolve(fileURLToPath(import.meta.url));
if (DIRECT) process.exitCode = main(process.argv.slice(2));
